Type annotations improvements

En3Tho opened the pull request:
I've been working on this for some time and instead of writing tests right away I've used ide manual checking. So now all of those tests are gone it seems -.-

This was supposed to intoduce granular parameter annotation, let!/use! local values annotations, member and property annotations, logically split function annotation and parameter/value annotation, also fix functions with no parameters but without tests it's hard to check.. I will write some more when I have more time.

Problems with my code:
Initially I took approach of splitting every annotateable pattern into separate functions but this approach is problematic because I coudn't find any way to get type of ArrayOrListPat/WildPat/ActivePatternPat (or something like that) on it's own. I mean it looks like there is simply no corresponding unique fcssymbol with type information for those.

That made me split my code into 2 parts - handling parameters and non-parameters. I think this is a kinda working approach for now but I'm not happy with my implementation because it's messy right now.

The idea is that we either have locals or parameters and for parameters we can extract the type because there is a function or a method providing those. So even aforementioned problematic patterns are easy to annotate.

For locals it's different. We can annotate IReferencePatterns only, sadly.

I'm sorta struggling making the right split for patterns with types specified and patterns with no types. I guess some kind of option is okay?

Changes are mostly moving tests to different folders

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/MemberAnnotationAction.fs:
```diff
@@ -0,0 +1,56 @@
+
+        let symbolUse =
+            let checker = memberDeclaration.FSharpFile.FcsCheckerService
+            checker.ResolveNameAtLocation(memberDeclaration, [| memberDeclaration.DeclaredName |], true, "Get declaration")
```
Please use `SourceName` in places like this. `DeclaredName` is how R# sees the name in cross-language analysis and this name may be a different compiled name.

In this particular place `IFSharpDeclaration.GetFcsSymbol` (or `-SymbolUse`) is what should be used instead.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/MemberAnnotationAction.fs:
```diff
@@ -0,0 +1,56 @@
+                SpecifyTypes.specifyPropertyType displayContext mfv.ReturnParameter.Type memberDeclaration
+            | _ ->
+                let types = FcsTypeUtil.getFunctionTypeArgs false mfv.FullType
+                (memberDeclaration.ParametersDeclarations, types)
```
Likewise, the `Enumerable` variant is preferred.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/ToRecursiveLetBindingsAction.fs:
```diff
@@ -37,4 +39,4 @@ type ToRecursiveLetBindingsAction(dataProvider: FSharpContextActionDataProvider)
         ToRecursiveLetBindingsAction.Execute(letBindings)
 
-        null
+        null
```
Please don't remove unix-compliant newlines at EOF 🙂

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/TupleAnnotationAction.fs:
```diff
@@ -0,0 +1,26 @@
+
+[<ContextAction(Name = "AnnotateBinding", Group = "F#",
+                Description = "Annotate binding with explicit type")>]
+type TupleAnnotationAction(dataProvider: FSharpContextActionDataProvider) =
```
It seems there should be no separate action for a specific pattern kind. I think there should be an action for a simple reference pattern and it should probably also provided a scoped action for the whole containing compound pattern to cover cases like this.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/ValueAnnotationAction.fs:
```diff
@@ -0,0 +1,44 @@
+        let localReference = dataProvider.GetSelectedElement<ILocalReferencePat>()
+
+        isNotNull localReference
+        &&  match localReference.Parent with
```
Please move `&&` to the previous line.

auduchinok reviewed (commented):
AFAIK this PR is going to change considerable, so I left just a few generic comments.

En3Tho commented:
Not sure build failure is related

En3Tho commented:
This is ready for review.

En3Tho commented:
I've resolved all previous review comments I believe. This needs a new review anyway.

En3Tho commented:
Added an exception for ConstrainedTypeUsage - that was sorta quirky

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
```
Getting next sibling may not work here, after skipping parens and getting an inner type usage above.

auduchinok replied:
Imagine a case like this:
```fsharp
let (Some x): _ option = None
```
If you start with the paren pattern and get its inner pattern afterwards, it won't be able to get the next sibling of the paren pattern. So maybe you should use the inner parent for checking if it's annotated and the outer for getting the next sibling. But, anyway, getting the next sibling manually is not as good as using the generated return type accessor on `IBinding`.

En3Tho replied:
I'm getting this from BindingNavigator so it should be safe I guess

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+        | :? ITypedPat as typedPat ->
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
```
When is this possible?

En3Tho replied:
Like in example above, let x: _ list = []

En3Tho replied:
Removed this thing.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+module AnnotationUtil =
+
+    let private isFullyAnnotatedNamedTypeUsage (namedTypeUsage: INamedTypeUsage) =
+        isNull namedTypeUsage.ReferenceName ||
```
Not critical: this looks a bit counter-intuitive that not having a name makes it considered as fully annotated. Maybe a name like `needsAnnotation` could make it clearer.

En3Tho replied:
I've changed this a little, please take a look

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+
+    let isFullyAnnotatedBinding (binding: IBinding) =
+        isNotNull binding.ReturnTypeInfo &&
+        binding.ParametersDeclarationsEnumerable
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
```
```suggestion
        binding.ParameterPatternsEnumerable |> Seq.forall isFullyAnnotatedPattern
```

or even

```suggestion
        Seq.forall isFullyAnnotatedPattern binding.ParameterPatternsEnumerable
```

En3Tho replied:
I would rather use first suggestion if you don't mind.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
+
+    let isFullyAnnotatedBinding (binding: IBinding) =
+        isNotNull binding.ReturnTypeInfo &&
```
The return type info is not null here:
```fsharp
let a: _ * int = 1, 2
let b: [<A>] _ = 1
```

En3Tho replied:
Yes, my bad. I was meant to use isFullyAnnotatedReturnTypeInfo here. This shows that I don't have enough tests it seems.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isNotNull memberDeclaration.ReturnTypeInfo &&
```
Maybe this check could be moved into `isFullyAnnotatedReturnTypeInfo` to be consistent with `isFullyAnnotatedNamedTypeUsage`.

En3Tho replied:
Moved

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isNotNull memberDeclaration.ReturnTypeInfo &&
+        isFullyAnnotatedReturnTypeInfo memberDeclaration.ReturnTypeInfo &&
+        memberDeclaration.ParametersDeclarations
```
Please use `ParameterPatternsEnumerable`, it'll simplify things here.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+    // let f<'a> = ...
+    let isFunctionBinding (binding: IBinding) =
+        binding.HasParameters
+        || binding.HeadPattern.GetNextMeaningfulSibling() :? IPostfixTypeParameterDeclarationList
```
I'd not expect a generic value to be considered as a function from F# point of view. Why is this implemented like this?

En3Tho replied:
Well, from my point of view it behaves exactly like a function. E.g. it let f<'a> = ... behaves like an let f<'a>() = ...

auduchinok replied:
It's indeed compiled to a method but it's not considered a function from F# point of view, i.e. it doesn't have a function type.

En3Tho replied:
Should I move it to value annotation action?

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+        if forceSpaceBeforeColon || pattern.GetSettingsStoreWithEditorConfig().GetValue(fun (key: FSharpFormatSettingsKey) -> key.SpaceBeforeColon) then
+            ModificationUtil.AddChildBefore(pattern, Whitespace()) |> ignore
+
+    let private addReturnTypeInfo typeString (node: ITreeNode) =
+        let factory = node.CreateElementFactory()
+        let typeUsage = factory.CreateTypeUsage(typeString)
+        ModificationUtil.AddChildAfter(node, factory.CreateReturnTypeInfo(typeUsage))
```
I wonder if using generated `IBindingLikeDeclaration.SetReturnTypeInfo` would produce the expected result here instead?

Even if it doesn't work for some reason, we shouldn't probably add a function that can add a return type info to any node type.

En3Tho replied:
We found that this api most probably doesn't work now. I've removed this function and inlined implementations into specifyBinding/Method/PropReturnType

auduchinok reviewed (commented):
@En3Tho I only went through a small top part of the changes today. The core idea (in these changes) seems correct. In some cases there're safer ways to navigate the tree that would also make it easier to change the tree and grammar later, so I left some comments.

I'm going to continue the review in a day or two.

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+
+module Declaration =
+    [<return: Struct>]
+    let (|IsNotNullAndHasMfvSymbolUse|_|) (declaration: IFSharpDeclaration) =
+        if isNull declaration then ValueNone else
+        let symbolUse = declaration.GetFcsSymbolUse()
+        if isNull symbolUse then ValueNone else
+
+        match symbolUse.Symbol with
+        | :? FSharpMemberOrFunctionOrValue as mfv ->
+            ValueSome struct (symbolUse, mfv)
+        | _ ->
+            ValueNone
```
It's just to make this check more convinient than multi-line nullchecks. I can remove it if you wish.

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+        | :? ITypedPat as typedPat ->
+            ModificationUtil.ReplaceChild(typedPat, typedPat.Pattern.IgnoreInnerParens())
+        | _ ->
+            match getTypeUsage pattern with
+            | ValueSome typeUsage ->
+                ModificationUtil.DeleteChild(typeUsage)
```
let x = _ list for example

auduchinok replied:
The type usage is placed into a separate return type info node and is not a sibling of the pattern here.

[PSI structure](https://gist.github.com/auduchinok/077431520863efb2121f3a1de48259b1)

En3Tho replied:
Removed this as not needed.

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+        let typeUsage = factory.CreateTypeUsage(typeString)
+        ModificationUtil.AddChildAfter(node, factory.CreateReturnTypeInfo(typeUsage))
+
+    let private replaceWithTypedPattern typeString (pattern: IFSharpPattern) =
```
Well, it's just for readability. I belive it's easier to read/maintain.
I can fix it

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+
+        find node
+
+    let getTypeUsage (pattern: IFSharpPattern) =
+        findSibling<ITypeUsage> pattern
```
removed this function as it's actually not needed.

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+            ModificationUtil.ReplaceChild(pattern, updatedPattern)
+
+module FcsMfvUtil =
+    let getFunctionReturnType parameters (fcsType: FSharpType) =
```
Moved functions there but there is still GetFunctionType function exist and I don't really like it. I would rather use something less allocating

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+
+module FcsMfvUtil =
+    let getFunctionReturnType parameters (fcsType: FSharpType) =
+        let rec skipFunctionParameters remaining (fcsType: FSharpType) =
+            if remaining = 0 then fcsType
+            else
+                skipFunctionParameters (remaining - 1) fcsType.GenericArguments[1]
```
renamed to skipInnerTypes

En3Tho commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,364 @@
+        parenPat.SetPattern(pattern) |> ignore
+        ModificationUtil.ReplaceChild(pattern, parenPat) |> ignore
+
+    let private specifyConstrainedTypeUsage displayContext (fcsType: FSharpType) (typeUsage: IConstrainedTypeUsage) =
```
Can you please tell me if it's required to be inlined or is this a question? If it is then I would like to keep smaller functions just for the sake of readability. If it is required then I will fix it of course.

En3Tho commented:
I did some fixes, but not all of them. I guess more tests actually need to be written (like those examples you've shown in CR).
I like this generic way of annotating things but I guess to rewrite it I need more stuff to check.

En3Tho commented:
@auduchinok I belive I've fixed all the thing you noted except for those that are still in question. I've added those tricky stuff you've mentioned in tests like let x: [] _ = 1
Please look at specifyReturnTypeInfo function, it is super quircky, sadly.
Should I add more tests or maybe remove some?
And I guess I have to run other tests to see if I changed behavior of some other features.

En3Tho commented:
@auduchinok Any new input about this pr?

auduchinok commented:
@En3Tho Sorry for the delay, I'll try to get back to it today.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/ValueAnnotationAction.fs:
```diff
@@ -0,0 +1,119 @@
+        // not sure if member declaration can return null symbolUsage
+        // if this is impossible then moving actual transition is won't be too hard
+
+        isAvailableForParameter()
+        || isAvailableForValue()
```
This would probably be simpler as two separate actions, as it won't require neither adding a special type to distinguish the two cases, not matching it and adding a suspicious assert.

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/ValueAnnotationAction.fs:
```diff
@@ -0,0 +1,119 @@
+        | Declaration.IsNotNullAndHasMfvSymbolUse(symbolUse, mfv) ->
+            let parameterIndex = parameters |> Seq.findIndex (fun bindingParam -> bindingParam == parameter)
+            let parameterIndex = if isFromMember then parameterIndex + 1 else parameterIndex
+            let actualParameterType = FcsTypeUtil.getFunctionParameterAt parameterIndex mfv.FullType
```
This throws exceptions on the following code:

```fsharp
type T =
    static member M(a, b) = ()
```

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/MemberAnnotationAction.fs:
```diff
@@ -0,0 +1,62 @@
+                SpecifyUtil.specifyPropertyType displayContext mfv.ReturnParameter.Type memberDeclaration
+            | _ ->
+                let parameters = memberDeclaration.ParameterPatterns
+                let types = FcsTypeUtil.getMethodParameterTypes parameters.Count mfv.FullType
```
This throws exceptions on the following code:

```fsharp
type T =
    static member M(a) = ()
```

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs:
```diff
@@ -0,0 +1,332 @@
+        let newTypeInfo = factory.CreateReturnTypeInfo(factory.CreateTypeUsage(typeString))
+
+        let mutable idx = 0
+        if not (returnTypeInfo.AttributeListsEnumerable.IsEmpty()) then
+            for attributeList in returnTypeInfo.AttributeListsEnumerable do
+                FSharpAttributesUtil.addAttributeList false newTypeInfo.ReturnType
+                let mutable prevAttribute = null
+                for attribute in attributeList.AttributesEnumerable do
+                    if isNull prevAttribute then
+                        prevAttribute <- FSharpAttributesUtil.addAttribute newTypeInfo.AttributeLists[idx] attribute
+                    else
+                        FSharpAttributesUtil.addAttributeAfter prevAttribute attribute
+                idx <- idx + 1
```
Why does it need replacing the whole return type info instead of replacing the type?

auduchinok commented on ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/ValueAnnotationAction.fs:
```diff
@@ -0,0 +1,119 @@
+
+        let pattern =
+            match pattern.Parent with
+            | :? ITuplePat as tuplePat ->
+                tuplePat :> IFSharpPattern
```
`ParenPatUtil.addParensIfNeeded` may help here.

auduchinok commented:
@En3Tho Thanks for bringing this up. I'm going to work on it. I think the tests could be useful indeed, thanks!
